addons/account/models/account_move.py

```diff
@@ -991,29 +998,36 @@ def _recompute_dynamic_lines(self, recompute_all_taxes=False, recompute_tax_base
         '''
         for invoice in self:
             # Dispatch lines and pre-compute some aggregated values like taxes.
+            expected_tax_rep_lines = set()
+            current_tax_rep_lines = set()
+            for line in invoice.line_ids:
+                if line.recompute_tax_line:
+                    recompute_all_taxes = True
+                    line.recompute_tax_line = False
+                if line.tax_repartition_line_id:
+                    current_tax_rep_lines.add(line.tax_repartition_line_id._origin)
+                elif line.tax_ids:
+                    if self.is_invoice(include_receipts=True):
+                        is_refund = self.type in ('out_refund', 'in_refund')
+                    else:
+                        tax_type = line.tax_ids[0].type_tax_use
+                        is_refund = (tax_type == 'sale' and line.debit) or (tax_type == 'purchase' and line.credit)
+                    taxes = line.tax_ids.flatten_taxes_hierarchy()
+                    if is_refund:
+                        tax_rep_lines = taxes.refund_repartition_line_ids._origin.filtered(lambda x: x.repartition_type == "tax")
+                    else:
+                        tax_rep_lines = taxes.invoice_repartition_line_ids._origin.filtered(lambda x: x.repartition_type == "tax")
+                    for tax_rep_line in tax_rep_lines:
+                        expected_tax_rep_lines.add(tax_rep_line)
+            delta_tax_rep_lines = expected_tax_rep_lines - current_tax_rep_lines
 
-            # determine required tax_lines by finding all non-zero tax amounts
-            lines_with_taxes = invoice.line_ids.filtered(
-                lambda l: l.tax_ids and l.move_id.currency_id.compare_amounts(l.price_subtotal, l.price_total))
-
-            needed_taxes = lines_with_taxes.tax_ids._origin.flatten_taxes_hierarchy().filtered(
-                lambda tax: (
-                        tax.amount_type == 'fixed' and not invoice.company_id.currency_id.is_zero(tax.amount)
-                        or not float_is_zero(tax.amount, precision_digits=4)
-                )
-            )
-            current_taxes = invoice.line_ids.tax_line_id._origin
-
-            if (
-                recompute_all_taxes
-                or any(line.recompute_tax_line for line in invoice.line_ids)
-                or needed_taxes > current_taxes
-            ):
-                invoice.line_ids.recompute_tax_line = False
+            # Compute taxes.
+            if recompute_all_taxes:
                 invoice._recompute_tax_lines()
-
-            if recompute_tax_base_amount:
+            elif recompute_tax_base_amount:
                 invoice._recompute_tax_lines(recompute_tax_base_amount=True)
+            elif delta_tax_rep_lines and not self._context.get('move_reverse_cancel'):
+                invoice._recompute_tax_lines(tax_rep_lines_to_recompute=delta_tax_rep_lines)
```

Comment:
What if `recompute_tax_base_amount=True` but there are repartitions lines missing?

Reply:
Idk but the `recompute_tax_base_amount` is already hacky at the first place. I didn't want to change the current behavior